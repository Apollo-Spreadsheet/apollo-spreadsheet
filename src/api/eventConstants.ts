export const GRID_FOCUS_OUT = 'GRID_FOCUS_OUT'
export const ROW_SELECTION_CHANGE = 'ROW_SELECTION_CHANGE'
export const CELL_CLICK = 'CELL_CLICK'
export const CELL_DOUBLE_CLICK = 'CELL_DOUBLE_CLICK'
export const CELL_BEGIN_EDITING = 'CELL_BEGIN_EDITING'
export const CELL_STOP_EDITING = 'CELL_STOP_EDITING'
export const DATA_CHANGED = 'DATA_CHANGED'
export const ROWS_CHANGED = 'ROWS_CHANGED'
export const COLUMNS_CHANGED = 'COLUMNS_CHANGED'