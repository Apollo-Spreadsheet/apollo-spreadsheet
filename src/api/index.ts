export * from './useApiFactory'
export * from './useApiEventHandler'
export * from './useApiExtends'
export * from './useApiRef'
export * from './eventConstants'
export * from './types'
