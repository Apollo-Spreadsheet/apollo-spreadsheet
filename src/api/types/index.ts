export * from './apiRef'
export * from './coreApi'
export * from './editorManagerApi'
export * from './gridApi'
export * from './mergeCellsApi'
export * from './navigationApi'
export * from './rowsApi'
export * from './rowSelectionApi'
export * from './nestedRowsApi'
export * from './themeApi'
