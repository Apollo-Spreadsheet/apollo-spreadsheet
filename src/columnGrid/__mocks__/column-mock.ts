import { Header } from '../types/header.type'

export const createColumnMock = (column?: Partial<Header>) => {
	return {
		id: 'column - ' + Math.random(),
		accessor: column?.accessor ?? 'access-' + Math.random(),
		...column,
	} as Header
}
