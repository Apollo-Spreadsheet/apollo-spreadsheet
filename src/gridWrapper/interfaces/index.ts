export * from './getColumnWidthFn'
export * from './gridCell'
export * from './registerChildFn'