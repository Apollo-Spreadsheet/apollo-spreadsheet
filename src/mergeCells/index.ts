export * from './createMergedGroups'
export * from './createMergedPositions'
export * from './useMergeCells'
export * from './validateMergeData'
export * from './validationErrorMessages'
export * from './validators'
export * from './interfaces'
