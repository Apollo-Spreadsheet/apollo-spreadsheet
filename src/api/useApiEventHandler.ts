import { useEffect } from 'react'
import { ApiRef } from './types/apiRef'
import { useLogger } from '../logger'

export function useApiEventHandler(
	apiRef: ApiRef,
	eventName: string,
	handler?: (args: any) => void,
) {
	const logger = useLogger('useApiEventHandler')
	useEffect(() => {
		if (handler && eventName) {
			logger.debug(`Subscribing to ${eventName} with handler reference ${handler.name}`)
			return apiRef.current.subscribeEvent(eventName, handler)
		}

		return undefined
	}, [apiRef, eventName, handler, logger])
}
