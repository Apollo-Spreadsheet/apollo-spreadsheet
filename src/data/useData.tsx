import React, { useMemo } from 'react'
import { Header } from '../columnGrid/types/header.type'
import { MergeCell } from '../mergeCells/interfaces/merge-cell'
import { Checkbox } from '@material-ui/core'
import { SelectionProps } from '../rowSelection/selectionProps'
import { MergePosition } from '../mergeCells/createMergedPositions'
import { GridCell } from '../gridWrapper/interfaces/gridCell'
import { insertDummyCells } from '../gridWrapper/utils/insertDummyCells'
import { formatCellValue } from './formatCellValue'

interface Props<Row> {
	rows: Row[]
	headers: Header[]
	mergeCells?: MergeCell[]
	mergedPositions?: MergePosition[]
	isRowSelected: (rowId: string) => boolean
	selectRow: (rowId: string) => void
	selection?: SelectionProps<Row>
}

/**
 * useData transforms the given data/rows into cells for internal usage and validates but also provides utilities
 */
export function useData<Row = any>({
	rows,
	headers,
	mergeCells,
	mergedPositions,
	isRowSelected,
	selectRow,
	selection,
}: Props<Row>) {
	const data = useMemo(() => {
		const cellsList = rows.reduce((list, row, i) => {
			const cells = headers.reduce((_cells, header, colIndex) => {
				const isDummy = mergedPositions?.some(e => e.row === i && e.col === colIndex)
				if (isDummy) {
					return _cells
				}

				const mergeInfo = mergeCells?.find(e => e.rowIndex === i && e.colIndex === colIndex)
				const cellValue = row[header.accessor] !== undefined ? row[header.accessor] : ''
				const value = header.cellRenderer
					? (header.cellRenderer({ row, column: header }) as any)
					: formatCellValue(cellValue)

				_cells.push({
					colSpan: mergeInfo?.colSpan,
					rowSpan: mergeInfo?.rowSpan,
					value: value,
				})
				return _cells
			}, [] as GridCell[])

			if (selection) {
				const isSelectable = selection.canSelect ? selection.canSelect(row) : true
				if (isSelectable) {
					const selected = isRowSelected(row[selection.key])
					//Bind the rowSelection
					cells[cells.length - 1] = {
						value: (
							<Checkbox
								className={selection.checkboxClass}
								checked={selected}
								onClick={() => selectRow(row[selection.key])}
							/>
						),
					}
				}
			}

			list[i] = cells
			return list
		}, [] as GridCell[][])

		return insertDummyCells(cellsList) as GridCell[][]
	}, [rows, headers, mergeCells, isRowSelected, selectRow, selection])

	return { data }
}
