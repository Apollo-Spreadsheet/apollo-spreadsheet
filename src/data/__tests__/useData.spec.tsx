import React, { useRef } from 'react'
import { renderHook, act } from '@testing-library/react-hooks'
import { useData } from '../useData'
import { createColumnMock } from '../../columnGrid/__mocks__/column-mock'
import { MergeCell } from '../../mergeCells/interfaces'
import { useMergeCells } from '../../mergeCells'
import { useApiFactory, useApiRef } from '../../api'

describe('useData hook', () => {
	const { result: { current: apiRefMock }} = renderHook(() => {
		const ref = useApiRef()
		const divRef = useRef(document.createElement('div'))
		useApiFactory(divRef, ref)
		return ref
	})
	it('should mount without any data', () => {
		const { result } = renderHook(() => useData({
			apiRef: apiRefMock,
			rows: [],
			columns: [],
			initialised: apiRefMock.current.isInitialised
		}))
		expect(result.current.rows.length).toBe(0)
		expect(result.current.cells.length).toBe(0)
	})

	it('should mount with rows (no merge cells)', () => {
		const mockHeaders = [createColumnMock()]
		const rows = [{ id: 1 }]
		const { result } = renderHook(() => useData({
			rows,
			columns: mockHeaders,
			apiRef: apiRefMock,
			initialised: apiRefMock.current.isInitialised
		}))
		expect(result.current.rows).toEqual(rows)
		expect(result.current.cells).toEqual([[{
			colSpan: undefined,
			rowSpan: undefined,
			value: '',
		}]])
	})

	it('should mount with rows and merged cells', () => {
		const mockHeaders = [createColumnMock()]
		const rows = [{ id: 1 }, { id: 2 }, { id: 3 }]
		const mergeCells: MergeCell[] = [{
			rowIndex: 0,
			rowSpan: 2,
			colSpan: 1,
			colIndex: 0,
		}]
		act(() => {
			renderHook(() => {
				useMergeCells({
					apiRef: apiRefMock,
					mergeCells,
					columnCount: mockHeaders.length,
					rowCount: rows.length,
					initialised: true
				})
			})
		})
		const { result } = renderHook(() => useData({
			rows,
			columns: mockHeaders,
			apiRef: apiRefMock,
			initialised: apiRefMock.current.isInitialised
		}))

		expect(result.current.cells).toMatchSnapshot()
	})
})