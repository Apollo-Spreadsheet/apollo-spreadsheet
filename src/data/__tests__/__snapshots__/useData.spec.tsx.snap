// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useData hook should bind selection cell 1`] = `
Array [
  Array [
    Object {
      "value": <ForwardRef(WithStyles)
        onClick={[Function]}
      />,
    },
  ],
]
`;

exports[`useData hook should mount with rows and merged cells 1`] = `
Array [
  Array [
    Object {
      "colSpan": 1,
      "rowSpan": 2,
      "value": "",
    },
  ],
  Array [
    Object {
      "colSpan": undefined,
      "dummy": true,
      "dummyFor": "rowSpan",
      "first": true,
      "last": true,
      "parentRow": Object {
        "rowSpan": 2,
      },
      "rowSpan": 1,
      "style": Object {
        "borderTop": 0,
      },
      "value": "",
    },
  ],
  Array [
    Object {
      "colSpan": undefined,
      "rowSpan": undefined,
      "value": "",
    },
  ],
]
`;
